skfda/clustering.py

```diff
@@ -0,0 +1,1180 @@
+    def p(self):
+        return self._p
+
+    @property
+    def centers(self):
+        return self._centers
+
+    @property
+    def n_iter(self):
+        return self._n_iter
+
+    def _generic_clustering_checks(self, fdatagrid, n_clusters, init):
+        """Checks the arguments *fdatagrid*, *n_clusters* and *init*
+        used in the :func:`fit <fda.clustering.ClusteringData.fit>` function.
+
+        Args:
+            fdatagrid (FDataGrid object): Object whose samples
+                are classified into different groups.
+            n_clusters (int): Number of groups into which the samples are
+                classified.
+            init (ndarray): Contains the initial centers of the different
+                clusters the algorithm starts with or None.
+
+        Returns:
+            fdatagrid (FDataGrid object): Object whose samples
+                are classified into different groups.
+            n_clusters (int): Number of groups into which the samples are
+                classified.
+            init (ndarray): Contains the initial centers of the different
+                clusters the algorithm starts with. If None, default value,
+                the centers are generated randomly.
+        """
+
+        if fdatagrid.ndim_domain > 1:
+            raise NotImplementedError(
+                "Only support 1 dimension on the domain.")
+
+        if fdatagrid.nsamples < 2:
+            raise ValueError(
+                "The number of observations must be greater than 1.")
+
+        if n_clusters < 2:
+            raise ValueError(
+                "The number of clusters must be greater than 1.")
+
+        if init is not None and init.shape != (
+                fdatagrid.ndim_image, n_clusters, fdatagrid.ncol):
+            raise ValueError(
+                "The init ndarray should be of shape (ndim_image, n_clusters, n_features) "
+                "and gives the initial centers.")
+        elif init is None:
+            init = np.array([None] * fdatagrid.ndim_image)
+
+        return fdatagrid, n_clusters, init
+
+    def _random_initialization_centers(self, data_matrix, fdatagrid,
+                                       n_clusters):
+        """Random initialization of the centroids used in the
+        :func:`fit function <fda.clustering.ClusteringData.fit>` if *init*
+        is None.
+
+        Args:
+            data_matrix (ndarray): matrix with the data only of the
+                dimension of the image of the fdatagrid the algorithm is
+                classifying.
+            fdatagrid (FDataGrid object): Object whose samples are
+                classified into different groups.
+            n_clusters (int): Number of groups into which the samples are
+                classified.
+
+        Returns:
+            centers (ndarray): initial centers
+        """
+        comparison = True
+        random_state_aux = self.random_state
+        while comparison:
+            random.seed(a=random_state_aux)
```

Comment:
Do NOT change global state in a library.

Reply:
I am using the random module to calculate the initial centroids randomly. If I used numpy random module would it be fine? The code would be:
```
    def _random_initialization_centers(self, data_matrix, fdatagrid,  n_clusters):
        comparison = True
        random_state_aux = self.random_state
        while comparison:
            np.random.seed(a=random_state_aux)
            indices = np.random.choice(fdatagrid.nsamples, n_clusters, replace=False)
            centers = data_matrix[indices]
            comparison = np.asarray([(centers[i] == centers).all()
                                     for i in range(n_clusters)]).sum()
            random_state_aux += 1

        return centers
```
If we want to obtain the same centroids, that is, reproduce the same result, the random seed must be customized since the numpy.random functions do not accept the seed as a parameter.

Reply:
No, the key is making a RandomState object, which has the necessary methods (https://docs.scipy.org/doc/numpy/reference/generated/numpy.random.RandomState.html). You can use as an example the code for the synthetic dataset generation: https://github.com/GAA-UAM/scikit-fda/blob/a7f32d156468b5ff21f70565be5fd9bcbab68ee2/skfda/datasets/_samples_generators.py#L35-L54

Reply:
https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_random_state.html It says that it returns a new RandomState that it is what I am doing.

Reply:
The user should be able to provide a seed, a RandomState object or None. That is why we use `sklearn.utils.check_random_state`.